code/__HELPERS/turfs.dm

```diff
@@ -305,6 +305,72 @@ Turf and target are separate in case you want to teleport some distance from a t
 
 	return turf_list
 
+/// returns a list of turf lists.
+/// each list has turfs from every impact value
+/// * turf_lists[1] = list(1 turfs)
+/// * turf_lists[2] = list(8 turfs) (edge of 2nd impact)
+/// * turf_lists[3] = list(16 turfs) (edge of 3rd impact)
+/// * turf_lists[4] = list(24 turfs) (edge of 4th impact)
+/// NOTE: this doesn't support calculating z turfs yet.
+/proc/get_pulsing_turfs(turf/center, impact_size, check_ticks = TRUE)
```

Comment:
This proc kinda sucks, its way better to get the sides of the radius when you need them rather than try to get all of them at once and deal with lists. You also don't need to, or want to, be dealing with |= for appending 2 large lists together since it requires n*m operations instead of just m. You don't need to use |= since if the radius is r, then you can go from:
(-r, -r) to (r, -r)
(-r, -r + 1) to (-r, r - 1)
(-r, r) to (r, r)
(r, r - 1) to (r, -r + 1)

If you want a function that can get all these for convenience, then it should be an enumerator datum and should lazilly generate the turfs as it enumerates through itself.

Reply:
(Copypasta what I worte in Discord)
calculating means running the actual effect is delayed by calculating
that's why I stored it
the old logic does "calculating -> does effects -> calculating -> does effects -> repeats...", and that's potentially slow 
so I wanted to calculate all turfs first, then does effects with stored turfs.

> You also don't need to, or want to, be dealing with |= for appending 2 large lists together since it requires n*m operations instead of just m. 

but also, separating turfs into pulse group list means the cost of |= is low
you still need to check if there's the same turf inside, but it's only checked within the same grouped list, which doesn't compare with the whole turfs.

Also, I tried mathematical algorithm rather than using block(), and I found out using while() or for() is incredibly expensive than using block() and checking individual pulse group.
So, I really can **strongly** say this **IS** best.

Reply:
also, making a code that doesn't use |= makes the code more complex and unreadable. It's really better.

Reply:
Its the difference between >300000 operations and 796 operations.
All it needs is something like this
```
/proc/get_ring(x, y, z, radius)
  var/list/selection = list()
  selection += block(locate(max(x-radius, 1), min(y+radius, world.maxy), z), locate(x+radius, min(y+radius, world.maxy), z))
  selection += block(locate(max(x-radius, 1), max(y-radius, 1), z), locate(min(x+radius, world.maxx), max(y-radius, 1), z))
  selection += block(locate(max(x-radius, 1), max(y-radius+1,1), z), locate(max(x-radius, 1), min(y+radius-1, world.maxy), z))
  selection += block(locate(min(x+radius, world.maxx), max(y-radius+1,1), z), locate(min(x+radius, world.maxx), min(y+radius-1, world.maxy), z))
  return selection
```

Reply:
![image](https://github.com/BeeStation/BeeStation-Hornet/assets/26465327/8665182a-69d1-4a88-af18-602e912ae5bd)
![image](https://github.com/BeeStation/BeeStation-Hornet/assets/26465327/77939ec5-7526-4781-8452-f877938f71b6)

Just to emphasise the speed difference. |= requires checking every single element in the list, which is an O(n) operation compared to an O(1) operation. This means that for a list of size n, if you want to insert m elements then you need n*m operations.